SeeShark/Disposable.cs

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) The Vignette Authors
+// This file is part of SeeShark.
+// SeeShark is licensed under the BSD 3-Clause License. See LICENSE for details.
+
+using System;
+
+namespace SeeShark
+{
+    public abstract class Disposable : IDisposable
+    {
+        public bool Disposed { get; private set; }
+
+        public Disposable()
+        {
+            Disposed = false;
+        }
+
+        private void dispose(bool disposing)
+        {
+            if (Disposed)
+                return;
+
+            if (disposing)
+                DisposeManaged();
+
+            FreeUnmanaged();
+
+            Disposed = true;
+        }
+
+        public void Dispose()
+        {
+            dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        protected abstract void DisposeManaged();
+        protected abstract void FreeUnmanaged();
```

Comment:
Keep the calling convention consistent. Call it `DisposeUnmanaged`.
```suggestion
        protected abstract void DisposeUnmanaged();
```

Reply:
This is actually intended to call it `FreeUnmanaged` and not `DisposeUnmanaged`. This function is supposed to handle the *freeing* (and not *disposal*) of unmanaged resources. The main difference is that we can dispose of a resource as much as we want, but we cannot free a resource more than once. Any object that implements `IDisposable` would actually be disposed in `DisposeManaged()`, and only unmanaged resources like raw pointers would be freed in `FreeUnmanaged()`.

Do you think it will still cause confusion given this nuance?

Reply:
the .NET convention regarding freeing up memory in the runtime is called object disposal, hence we use the `DisposeUnmanaged`, `Dispose` definitions instead. Please keep everything consistent as accordance to how the .NET classes are done please, don't do your own thing for this kinds of things.

Reply:
Ah don't worry, I don't want to do things my own way here. The whole reason I made this class was so that I could actually make sure that I follow the .NET standard. Thanks, I'm gonna look it up - maybe I misread stuff about disposal.

Reply:
`Dispose` instead of `Free` isn't that bad anyway, I just thought that it would have been nice to have this explicit distinction. But after some discussion, I decided that I was going to favor continuity over precise vocabulary.